### PR TITLE
Data covariance weighting in misfit calculation

### PR DESCRIPTION
**Pull request by emolch**

### CODE REVIEW

File: src/apps/grond.py [hunk 1]
```diff
@@ -919,18 +919,21 @@ def get_doc_title(doc):
     elif args[0] == 'all':
         if env is None:
             help_and_die(parser, 'two or three arguments required')
+        env.setup_modelling()
```
Comment (emolch) [inline src/apps/grond.py]: Not all plots need modelling. Why was this needed?

File: src/core.py [hunk 2]
```diff
@@ -141,6 +141,8 @@ def forward(rundir_or_config_path, event_names):
 def harvest(rundir, problem=None, nbest=10, force=False, weed=0):
 
     env = Environment([rundir])
+    env.setup_modelling()
```
Comment (emolch) [inline src/core.py]: Harvesting should not require modelling.

File: src/core.py [hunk 3]
```diff
@@ -164,7 +165,10 @@ def harvest(rundir, problem=None, nbest=10, force=False, weed=0):
 
     ibests_list = []
     ibests = []
-    gms = problem.combine_misfits(misfits)
+    gms = problem.combine_misfits(
```
Comment (emolch) [inline src/core.py]: As we now anyway store the combined misfits, it may be better to directly use these instead of combining them again.

Or we rethink if it was clever to allow the optimizer to intervene there. There is a conceptual difference between the `extra_weights` and the `extra_correlated_weights`. The latter do not affect the bootstrap chains differently while the former do. Correct? That would qualify the correlate weights to be part of the global objective function which should be completely defined in the problem.

Comment (emolch) [reply src/core.py]: Or both.

File: src/optimisers/highscore/optimiser.py [hunk 4]
```diff
@@ -15,6 +15,10 @@
 from grond.optimisers.base import Optimiser, OptimiserConfig, BadProblem, \
     OptimiserStatus
 
+import cProfile
```
Comment (emolch) [inline src/optimisers/highscore/optimiser.py]: Stray debug statements

File: src/optimisers/plot.py [hunk 5]
```diff
@@ -478,6 +491,7 @@ class BootstrapPlot(PlotConfig):
 
     name = 'bootstrap'
     size_cm = Tuple.T(2, Float.T(), default=(21., 14.9))
+    show_ticks = Bool.T(default=False)
```
Comment (emolch) [inline src/optimisers/plot.py]: Where is this used?

File: src/plot/main.py [hunk 6]
```diff
@@ -56,9 +56,12 @@ def make_plots(
     if plots_path is None:
         plots_path = env.get_plots_path()
 
+
     plots = plot_config_collection.plot_configs
     manager = PlotCollectionManager(plots_path, show=show)
+
     env.set_plot_collection_manager(manager)
+    env.setup_modelling()
```
Comment (emolch) [inline src/plot/main.py]: Should not call `setup_modelling` here.

File: src/problems/base.py [hunk 7]
```diff
@@ -38,6 +37,18 @@ def nextpow2(i):
     return 2**int(math.ceil(math.log(i)/math.log(2.)))
 
 
+def corr_misfits(w_misfits, weight_matrix):
+    ''' help function for the matrix multiplication
```
Comment (emolch) [inline src/problems/base.py]: Docstring does not conform to style guide.

File: src/problems/base.py [hunk 8]
```diff
@@ -278,8 +289,8 @@ def extract(self, xs, i):
     def get_target_weights(self):
         if self._target_weights is None:
             self._target_weights = num.concatenate(
-                [target.get_combined_weight()
-                 for target in self.targets])
+               [target.get_combined_weight()
```
Comment (emolch) [inline src/problems/base.py]: Unnecessary whitespace change.

File: src/problems/base.py [hunk 9]
```diff
@@ -388,71 +401,87 @@ def combine_misfits(
             to the residuals, indexed as
             ``extra_residuals[ibootstrap, iresidual]``.
 
+        :param extra_correlated_weights: if given a dictionary of
```
Comment (emolch) [inline src/problems/base.py]: * Describe how `idx_misfit` is to be interpreted.
* Maybe use `imisfit` to be consistent with `ibootstrap`, `iresidual`, etc.

File: src/problems/base.py [hunk 10]
```diff
@@ -388,71 +401,87 @@ def combine_misfits(
...
+            ``idx_misfit: correlated weight matrix`` has to be passed.
+
         :param get_contributions: get the weighted and perturbed contributions
             (don't do the sum).
 
-        :returns: if no *extra_weights* or *extra_residuals* are given, a 1D
+        :returns: if no *extra_weights* or *extra_residuals* or are given, a 1D
```
Comment (emolch) [inline src/problems/base.py]: Breaks the sentence.

File: src/problems/base.py [hunk 11]
```diff
@@ -388,71 +401,87 @@ def combine_misfits(
...
-                num.nansum(exp(w*(misfits[:, num.newaxis, :, 0]+r)), axis=2) /
-                num.nansum(exp(w*(misfits[:, num.newaxis, :, 1])), axis=2))
-            assert res[res < 0].size == 0
-            return res
-        else:
-            w = self.get_target_weights()[num.newaxis, :] \
-                * self.inter_family_weights2(misfits[:, :, 1])
+        if self.norm_exponent != 2 and extra_correlated_weights:
+            raise GrondError('Correlated weights can only be used '
+                             ' with norm_exponent=2!')
```
Comment (emolch) [inline src/problems/base.py]: I would prefer to not use `!` at the end of such messages. What is your opinion?

File: src/problems/base.py [hunk 12]
```diff
@@ -388,71 +401,87 @@ def combine_misfits(
...
+            weights = weights * extra_weights[num.newaxis, :, :]
+
+        for idx1, corr_weight_mat in extra_correlated_weights.items():
 
-            if get_contributions:
-                return exp(w*misfits[:, :, 0]) \
-                    / num.nansum(
-                        exp(w*misfits[:, :, 1]),
-                        axis=1)[:, num.newaxis]
+            idx2 = idx1 + corr_weight_mat.shape[0]
```
Comment (emolch) [inline src/problems/base.py]: `idx` -> `imisfit` if it is a misfit index.

Comment (emolch) [reply src/problems/base.py]: Otherwise `i`, `j` would be enough but we also have `imodel`, therefore I would prefer unambiguous names.

File: src/targets/base.py [hunk 13]
```diff
@@ -166,12 +172,11 @@ def get_bootstrap_residuals(self):
         return self.bootstrap_residuals.reshape(nbootstraps, self.nmisfits)
 
     def prepare_modelling(self, engine, source, targets):
-        return []
+        return [self]
```
Comment (emolch) [inline src/targets/base.py]: The base class implementation of `prepare_modelling` should return `[]`. There should be no default assumption that the target inherits from `pyrocko.gf.Target`.

File: src/targets/base.py [hunk 14]
```diff
@@ -166,12 +172,11 @@ def get_bootstrap_residuals(self):
...
 
     def finalize_modelling(
             self, engine, source, modelling_targets, modelling_results):
-
-        raise NotImplementedError('must be overloaded in subclass')
+        return modelling_results[0]
```
Comment (emolch) [inline src/targets/base.py]: same here.

File: src/targets/gnss_campaign/plot.py [hunk 15]
```diff
@@ -73,8 +73,12 @@ def draw_gnss_fits(self, ds, history, optimiser, vertical=False):
         gnss_targets = problem.gnss_targets
         for target in gnss_targets:
             target.set_dataset(ds)
+        print('aaa', history.misfits)
```
Comment (emolch) [inline src/targets/gnss_campaign/plot.py]: Stray prints.

File: src/targets/gnss_campaign/plot.py [hunk 16]
```diff
@@ -143,41 +147,33 @@ def plot_gnss(gnss_target, result, ifig, vertical=False):
...
 
-                m.add_gnss_campaign(model_camp, psxy_style={
-                    'G': 'red',
-                    'W': '0.8p,red',
-                    't': 30,
-                    },
-                    offset_scale=offset_scale, vertical=True, labels=False)
-            else:
-                m.add_gnss_campaign(campaign, psxy_style={
+            for ista, sta in enumerate(all_stations):
```
Comment (emolch) [inline src/targets/gnss_campaign/plot.py]: These changes do not seem to be related to the covariance weighting.

* Move into their own PR or to branch `v1.2.0-dev-misc`.

File: src/targets/gnss_campaign/target.py [hunk 17]
```diff
@@ -108,8 +112,15 @@ def string_id(self):
         return self.campaign_name
 
     def misfits_string_ids(self):
-        return ['%s.%s' % (self.path, station.code)
-                for station in self.campaign.stations]
+        id_list = []
```
Comment (emolch) [inline src/targets/gnss_campaign/target.py]: Some of these changes do not seem to be related to the covariance weighting.

* Move into their own PR or to branch `v1.2.0-dev-misc`.

File: src/targets/satellite/target.py [hunk 18]
```diff
@@ -220,13 +242,23 @@ def init_bootstrap_residuals(self, nbootstraps, rstate=None):
         scene = self.scene
         qt = scene.quadtree
         cov = scene.covariance
-        bootstraps = num.empty((nbootstraps, qt.nleaves))
+        bootstraps = num.zeros((nbootstraps, qt.nleaves))
+
+        # TODO:mi Signal handler is not given back to the main task!
+
+        with ThreadPoolExecutor(max_workers=os.cpu_count()) as executor:
```
Comment (emolch) [inline src/targets/satellite/target.py]: Might be good to limit the number of workers.

File: src/targets/waveform/target.py [hunk 19]
```diff
@@ -458,14 +458,6 @@ def post_process(self, engine, source, tr_syn):
             logger.debug(str(e))
             raise gf.SeismosizerError('No waveform data: %s' % str(e))
```
Comment (emolch) [inline src/targets/waveform/target.py]: See above.

**Comment by emolch** [comment]
TODO before merge:
* [ ] squash all commits belonging to covar

File: src/apps/grond.py (hunk 1 above)
Comment (miili) [reply src/apps/grond.py]: The `HighscoreOptimiser.get_correlated_weights` iterates the `problem.targets`, which need `target._ds` to retrieve the correlated weights.

File: src/core.py (hunk 2 above)
Comment (miili) [reply src/core.py]: Same here, the `extra_correlated_weights` need the dataset and target's weights.

File: src/optimisers/plot.py (hunk 5 above)
Comment (miili) [reply src/optimisers/plot.py]: This can be used in cusom plot configs but is not used in the standard output.

File: src/plot/main.py (hunk 6 above)
Comment (miili) [reply src/plot/main.py]: Same, we need the `extra_correlated_weights`

File: src/problems/base.py (hunk 11 above)
Comment (miili) [reply src/problems/base.py]: yes, rm

File: src/apps/grond.py (hunk 1 above)
Comment (emolch) [reply src/apps/grond.py]: Would it be possible to store these in the targets, so that they are part of `problem.yaml`?

Comment (miili) [reply src/apps/grond.py]: The matrices can become quite large ~ `800x800` for a single scene. I am hesitating to serialize that much data in the YAML.

File: src/core.py (hunk 3 above)
Comment (miili) [reply src/core.py]: Yes the misfits should not be combined again for plotting. Everything is stored in `bootstraps` anyways.

File: src/apps/grond.py (hunk 1 above)
Comment (miili) [reply src/apps/grond.py]: Resolved in https://github.com/pyrocko/grond/pull/33/commits/a02245a298b5619cb4849320d67084c82fc7e551

File: src/targets/satellite/target.py (hunk 18 above)
Comment (miili) [reply src/targets/satellite/target.py]: To less than the available cores?

File: src/targets/base.py (hunk 13 above)
Comment (emolch) [reply src/targets/base.py]: Ok, but check ToyTarget.

File: src/targets/base.py (hunk 14 above)
Comment (emolch) [reply src/targets/base.py]: Ok, but check ToyTarget.

File: src/targets/satellite/target.py (hunk 18 above)
Comment (emolch) [reply src/targets/satellite/target.py]: Grond should understand `--parallel=N` and `--threads=M`, and use these everywhere.

File: src/targets/base.py (hunk 13 above)
Comment (miili) [reply src/targets/base.py]: done.

File: src/problems/base.py [hunk 20]
```diff
@@ -466,13 +459,25 @@ def combine_misfits(
...
+        if num.any(extra_weights):
+            weights_tar = weights_tar * extra_weights[num.newaxis, :, :]
+
+        res = exp(res * weights_tar)
+        norms = exp(norms * weights_tar)
+ 
+        # get and apply normalization family weights (these weights depend on
+        # on just calculated norms!)
+        weights_fam = \
+            self.inter_family_weights2(root(norms[:, 0, :]))[:, num.newaxis, :]
```
[the remark this answers is not in the document]
Comment (miili) [reply src/problems/base.py]: Fixed in https://github.com/pyrocko/grond/pull/33/commits/da91a2c3908e1b3191bf0dd15320514b394dc600

[the remark this answers is not in the document]
Comment (hsudha) [reply src/problems/base.py]: No, I think, and it is not that simple. The normalization family weights (weights_fam) work is that the internal weighting in each target group does not change, but the external weight among other groups. This is why `weights_fam` have to be calculated after all the other weighting factors (in the `master` branch only simple products were involved so where the fam weights got drawn did not matter). Example: If there are three normalization families, they contribute equally to the global misfit. Test: put almost zero slip for a source and look at the contribution of targets. Each group should contribute approx 1/3 to the misfit, for two groups 1/2 etc.
BUT the way weights are set now, manual weights don't come into effect. Needs fixing.
